src/Val.cc

```diff
@@ -1023,34 +1072,69 @@ static std::variant<ValPtr, std::string> BuildVal(const rapidjson::Value& j, con
         }
 
         case TYPE_TABLE: {
-            if ( ! j.IsArray() )
-                return mismatch_err();
-
-            if ( ! t->IsSet() )
-                return util::fmt("tables are not supported");
-
-            auto tt = t->AsSetType();
-            auto tl = tt->GetIndices();
+            auto tt = t->AsTableType(); // The table vs set type does not matter below
             auto tv = make_intrusive<TableVal>(IntrusivePtr{NewRef{}, tt});
+            auto tl = tt->GetIndices();
 
-            for ( const auto& item : j.GetArray() ) {
-                std::variant<ValPtr, std::string> v;
+            if ( t->IsSet() ) {
+                if ( ! j.IsArray() )
+                    return mismatch_err();
 
-                if ( tl->GetTypes().size() == 1 )
-                    v = BuildVal(item, tl->GetPureType(), key_func);
-                else
-                    v = BuildVal(item, tl, key_func);
+                for ( const auto& item : j.GetArray() ) {
+                    std::variant<ValPtr, std::string> v;
 
-                if ( ! get_if<ValPtr>(&v) )
-                    return v;
+                    if ( tl->GetTypes().size() == 1 )
+                        v = BuildVal(item, tl->GetPureType(), key_func);
+                    else
+                        v = BuildVal(item, tl, key_func);
 
-                if ( ! std::get<ValPtr>(v) )
-                    continue;
+                    if ( ! get_if<ValPtr>(&v) )
+                        return v;
+                    if ( ! std::get<ValPtr>(v) )
+                        continue;
+
+                    tv->Assign(std::move(std::get<ValPtr>(v)), nullptr);
+                }
 
-                tv->Assign(std::move(std::get<ValPtr>(v)), nullptr);
+                return tv;
             }
+            else {
+                if ( ! j.IsObject() )
+                    return mismatch_err();
+
+                for ( auto it = j.MemberBegin(); it != j.MemberEnd(); ++it ) {
+                    rapidjson::Document idxstr;
+                    idxstr.Parse(it->name.GetString(), it->name.GetStringLength());
+
+                    std::variant<ValPtr, std::string> idx;
+
+                    if ( tl->GetTypes().size() > 1 )
+                        idx = BuildVal(idxstr, tl, key_func);
+                    else if ( IsQuotedJSONType(tl->GetPureType()) )
+                        // Parse this with the quotes the string came with. This
+                        // mirrors the quote-stripping in BuildJSON().
```

Comment:
Not sure this comment fits - there's not quotes in `it->name` IIUC for addr or enum. Maybe `IsQuotedJSONType()` could be `HasQuotesStripped()` or so.

Reply:
I've renamed the function to `UsesJSONStringType()`, perhaps that's better? I was trying to capture the list of Zeek types that use a JSON string for rendering.